### PR TITLE
Dither Plugin

### PR DESCRIPTION
**Pull request by clementroche**

### CODE REVIEW

**Comment by anthonyec** [comment]
There's a lot of commented out code. It'd be better to remove unused code since it avoids confusion about if the code should or shouldn't be there. And it's already tracked in git so won't be lost ;)

**Comment by clementroche** [comment]
done @anthonyec

File: package-lock.json [hunk 1]
```diff
@@ -17,18 +17,16 @@
         },
         "node_modules/@aashutoshrathi/word-wrap": {
             "version": "1.2.6",
-            "resolved": "https://registry.npmjs.org/@aashutoshrathi/word-wrap/-/word-wrap-1.2.6.tgz",
-            "integrity": "sha512-1Yjs2SvM8TflER/OD3cOjhWWOZb58A2t7wpE2S9XfBYTiIl+XFhQG2bjy4Pu1I+EAlCNUzRDYDdFwFYUKvXcIA==",
             "dev": true,
+            "license": "MIT",
```
Comment (niekert) [inline package-lock.json:21]: I wonder why the diff for `package-lock.json` is so huge.

Could you try resetting to main and running `npm install` again?
from the `plugins` repo root:

```
git fetch
git checkout origin/main package-lock.json
npm install
```

Then commit the diff and see if it makes more sense.

Comment (clementroche) [reply package-lock.json:21]: nothing changed, what npm version are you using ?

Comment (clementroche) [reply package-lock.json:21]: I added some packages:
- @radix-ui/react-slider
- colorjs.io"
- ogl
- clsx
- @rollup/rollup-linux-x64-gnu (build on vercel)

Comment (niekert) [reply package-lock.json:21]: Hmm weird, adding packages is fine but I think the diff is touching many unrelated packages. Using npm `10.8.2`

Maybe you should reset `package-lock.json` from upstream?

```
git remote add upstream git@github.com:framer/plugins.git
git fetch upstream

git checkout upstream/main package-lock.json
npm install
```

**Review by niekert — COMMENTED** [review]
Fantastic Plugin ❤️ Love it! I think it just needs the assert and package-lock.json to be fixed before we can merge

File: plugins/dither/src/App.tsx [hunk 2]
```diff
@@ -0,0 +1,213 @@
+import { ImageAsset, framer } from "framer-plugin"
+import { useCallback, useEffect, useRef, useState } from "react"
+import "./App.css"
+import { Renderer, Camera, Transform, Plane, Program, Mesh, Texture } from "ogl"
+import { OrderedDither } from "./materials/ordered"
+import cn from "clsx"
+
+import.meta.hot?.accept(() => {
```
[the remark this answers is not in the document]
Comment (clementroche) [reply plugins/dither/src/App.tsx:8]: it's required for development purpose

File: plugins/dither/src/App.tsx [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+
+        return () => cancelAnimationFrame(raf)
+    }, [render])
+
+    const saveImage = useCallback(async () => {
+        render()
+
+        const originalImage = await image.getData()
+
+        assert(gl.canvas)
```
[the remark this answers is not in the document]
Comment (clementroche) [reply plugins/dither/src/App.tsx:146]: nice catch

File: plugins/dither/src/App.tsx [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+        const start = performance.now()
+
+        await framer.setImage({
+            image: {
+                bytes: nextBytes,
+                mimeType: originalImage.mimeType,
+            },
+        })
+
+        console.log("total duration", performance.now() - start)
```
[the remark this answers is not in the document]
Comment (clementroche) [reply plugins/dither/src/App.tsx:159]: I'm leaving it on dev mode only such as:

```js
if (import.meta.env.DEV) {
  console.log("total duration", performance.now() - start)
}
```

File: package-lock.json (hunk 1 above)
Comment (clementroche) [reply package-lock.json]: is it better now ?

**Comment by clementroche** [comment]
can you check again please @niekert

File: package-lock.json (hunk 1 above)
Comment (niekert) [reply package-lock.json]: Yes 🙏  thanks!

**Review by niekert — APPROVED** [review]

**Comment by niekert** [comment]
@clementroche Unfortunately it won't let me merge because of conflicts with `main`. Mind rebasing the PR with upstream main and then we can merge? 🙏

**Pull request merged**